service/awsconfig/v5/resource/cloudformation/adapter/adapter.go

```diff
@@ -41,6 +41,8 @@ type Adapter struct {
 	MasterImageID    string
 	WorkerImageID    string
 
+	Outputs *outputsAdapter
```

Comment:
One problem with the current adapter implementation is separation of concerns. IMO there should be one adapter sorting one template. Having a clean data structure scoped for a dedicated purpose helps sorting this problem.

Reply:
The problem I see with this approach is that there are values that need to be used in more than one template. Currently, the data structure passed to the template parser is very simple, just flat, that allows us to use values where they are required. The adapter abstraction lets us group logic related to a single template, but there's no real restriction, we could have a single adapter or redistribute the resources in another layout organizing them as we see fit, the current approach would keep working. In this sense I think the templates are not separate concerns, some pieces of data need to be shared. 

So far we aimed to promote the common data field to the top level template, but haven't followed that approach strictly. IMO it would be better to keep sharing the logic between templates, if you want to have separate data structures for each template that's ok (although it adds a lot of code that isn't needed in fact), but the logic for common fields shouldn't be repeated.

Reply:
I do not think this is simple as it is right now. When we talked nobody could reliably tell which information is used where. That is a clear indicator that something is not right here. My suggestion here is to make one adapter responsible for one template to have a more clear data structure. Certain information are used in multiple places, right. So each adapter has to be configured with the information it needs in order to render its template properly. The problem I have with one flat structure for everything is the naming. You get names like `MasterSecurityGroupID` where it is way easier to work with `Master.SecurityGroup.ID` as a human. 

When you say we should not repeat logic for common fields, what is the problem with that? We have mostly key functions we use in multiple places and this is just fine IMO. I am coming from outside the project and suggest an improvement because the current implementation is not straight forward. I think it is just easier to understand and work with in case one adapter sorts one template. Otherwise I miss the point of using the adapters.

Reply:
> I do not think this is simple as it is right now. When we talked nobody could reliably tell which information is used where. That is a clear indicator that something is not right here

Yes, we had the convention to promote fields to the top-level adapter, but it wasn't followed. You are right it could be more clear, and what you propose is a possible solution, I agree. But I think that in this case it comes at a cost of flexibility. We see now separate templates, with related resources, but if you look at them closely some of the resources are associated with a concrete template but could be related to another. The current templates have evolved during development and we have shaped it as we found it more appropriate, but there could be a better arrangement.

During this development it has been very useful to have this flat data structure that let us move things around and try different layouts. IMO what you propose would involve more changes in code if things change in the arrangement, I hope that from now on there won't be too many changes though, so I'm ok with your approach, if we need to switch to a less restricted structure we can always do that in response to pains.

> When you say we should not repeat logic for common fields, what is the problem with that? We have mostly key functions we use in multiple places and this is just fine IMO

Totally, what I mean is that currently part of the code in the output adapter is repeated in the instance adapter. So I think that, as part of this PR, all that logic should be refactored, maybe in a follow up is ok too.

Reply:
There is for sure still work to be done. This here is not a complete refactoring and I expect the project in general to evolve heavily in the future. I am still exploring and try to get onto the same page. We also talked about the approach of cross functional teams vs component maintainers in our retro. I see that the consensus finding here is difficult for everyone of us. Nevertheless we should try to get on the same page so we know in which direction we can move towards to together. 

Is there now anything left I can do to get an approval on this PR? Is there anything missing from your point of view?